### ISSUE TITLE
Assert not null or specific object type for further assertions on an object

### ISSUE THREAD

**Issue by dave08** [kotest/kotest#601]
Say:
```kotlin
val obj: Person? = ....
obj shouldNotBeNullAnd {
    it.name shouldBe "b"
    it.age shouldNotBe 43
}
```
This should and assert that Person is not null, and provide the ability to check the non-null result. This could also probably be done with Kotlin Contracts, just by doing a null assertion before other assertions. The difference between this and `result!! should { }` is that the previous is an actual assertion and the latter is a runtime failure.

Something similar could possibly done for `shouldBeInstanceOf` to make assertions on the already casted instance.

**Comment by LeoColman** [comment]
I'm just a little bit afraid of using contracts as they might change a lot in the final version

**Comment by dave08** [comment]
I'm wondering if that's a big deal... since here we're talking about changing two or three functions. So even if they change, I'm doubtful they won't support the functionality, and those functions could be changed again... The tests on this feature shouldn't pass if they change and break the functionality.

**Comment by LeoColman** [comment]
I guess it shouldn't be a big deal, maybe you're correct.

I don't agree with the syntax tho. Do you think there's a better way to do it?

**Comment by dave08** [comment]
If contracts are used:

```kotlin
val obj: Person? = ....
obj.shouldNotBeNull()
// This should be already smart-casted
obj.name shouldBe "b"
obj.age shouldNotBe 43

val obj: Object = Person("b", 44)
obj.shouldBeInstance<Person>()
// This should be already smart-casted
obj.name shouldBe "b"
obj.age shouldNotBe 43

// or one could always do this using regular Kotlin features:
val obj: Person? = ....
obj.shouldNotBeNull().run {
// This should be already smart-casted
    name shouldBe "b"
    age shouldNotBe 43
}
```
The idea is to gain the assertion + smart-cast in one shot.

Otherwise, without contracts (or with the proposed contracts), what bothers you with the syntax?

**Comment by LeoColman** [comment]
Ok, so the thing with `ExperimentalContracts` is actually a big deal. If one use the `@ExperimentalContracts` annotation, every function that calls this function will have to use it too.

![image](https://user-images.githubusercontent.com/1577251/51473169-bd36da80-1d62-11e9-9051-de04fe7850f2.png)


I think it's a great idea, but I don't want to change user's code for an experimental feature, so this will be a big no-no for now.

**Comment by LeoColman** [comment]
I'll try to see if there's anyway to use this, but I don't think there is.

**Comment by dave08** [comment]
I wonder if just adding something like this to your gradle will solve this:

```groovy
tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
    kotlinOptions {
        freeCompilerArgs += [
                '-Xuse-experimental=kotlin.Experimental'
        ]
    }
}
```
@Kerooker

**Comment by LeoColman** [comment]
But would this solve it for your code, for example?

**Comment by LeoColman** [comment]
I think it's something to try. Let me see :)

**Comment by dave08** [comment]
Also see: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-use-experimental/index.html

> If a declaration is annotated with UseExperimental, its usages are not required to opt-in to that experimental API.

**Comment by LeoColman** [comment]
It still forces users to use the `@UseExperimental` in their classes, and that's not desired

**Comment by dave08** [comment]
Funny, because I'm sure there's quite a few libs using coroutine channels, so they also force the user to do this? I'll try to see what others say about this on Slack, maybe someone's used this more than us...

**Comment by LeoColman** [comment]
I'm trying to talk to people in Slack/General about this, but the only answer I've had so far is "This is impossible". But IDK

**Comment by dave08** [comment]
Here's another example and the install instructions don't indicate needing to specify `@UseExperimental`:

https://github.com/MarkusAmshove/Kluent/blob/396efe98ceb28003c00122214bbbaa3eac1b40ee/common/src/main/kotlin/org/amshove/kluent/Basic.kt

And:

https://github.com/MarkusAmshove/Kluent/pull/126

@Kerooker

**Comment by LeoColman** [comment]
Ok, it seems to be possible as another person from slack said. I'll try now

**Comment by LeoColman** [comment]
With the latest commit I added the possibility to use Kotlin Contracts for the type smart-cast.

However, I preferred to keep the consistency and not return any value from `shouldNotBeNull`. Do you think we still need a custom construct for `shouldNotBeNullAnd {}`, @sksamuel and @dave08 ?

**Comment by dave08** [comment]
Very nice @Kerooker, thanks! 

I personally don't think so, regular Kotlin constructs can be used for that. 

I suppose that `shouldBeInstanceOf<Foo>` won't even work with reified generics? Maybe for that it might be nice to have a lambda, it doesn't look like there's a target version for that feature. The only thing that bothers me is consistency. But it would certainly be nice.

Also I've had such need in collections too like here: https://github.com/robfletcher/strikt/commit/7c5251ced35fbec7ab7ec34448bb8b452fb52b50, but in KotlinTest, there's no need for `filter` since the tested collection is not wrapped, so really an assertion that a list contains only that type could be made, and then have a lamda to make further assertions on each value of the new casted collection.

**Comment by LeoColman** [comment]
`shouldBeInstanceOf<Foo>` is already reified. It's a currently unsupported feature :/

**Comment by dave08** [comment]
What do you think of the lambda in that case, or maybe that's too inconsistent?

**Comment by LeoColman** [comment]
I think it's inconsistent, but it wouldn't affect any current code. I don't really see why not, if it's well documented.

What do you think, @sksamuel ?

**Comment by LeoColman** [comment]
@dave08 
As this discussion won't be including Kotlin Contracts for now, please submit a feature request to add a lambda parameter to `shouldBeInstanceOf` for us to discuss it better.

For now I'll be closing this, as this feature was implemented and merged in the `3.3` branch.
